x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/datafeed/DelayedDataDetector.java

```diff
@@ -0,0 +1,144 @@
+            .map(bucket -> BucketWithMissingData.fromMissingAndBucket(calculateMissing(indexedData, bucket), bucket))
+            .collect(Collectors.toList());
+    }
+
+    private List<Bucket> checkBucketEvents(long start, long end) {
+        GetBucketsAction.Request request = new GetBucketsAction.Request(job.getId());
+        request.setStart(Long.toString(start));
+        request.setEnd(Long.toString(end));
+        request.setExcludeInterim(true);
+        request.setPageParams(new PageParams(0, (int)((end - start)/bucketSpan)));
+
+        try (ThreadContext.StoredContext ignore = stashWithOrigin(client.threadPool().getThreadContext(), ML_ORIGIN)) {
+            GetBucketsAction.Response response = client.execute(GetBucketsAction.INSTANCE, request).actionGet();
+            return response.getBuckets().results();
+        }
+    }
+
+    private Map<Long, Long> checkTrueData(long start, long end) {
+        String timeField = job.getDataDescription().getTimeField();
+        SearchSourceBuilder searchSourceBuilder = new SearchSourceBuilder()
+            .size(0)
+            .aggregation(new DateHistogramAggregationBuilder(DATE_BUCKETS).interval(bucketSpan).field(timeField))
+            .query(ExtractorUtils.wrapInTimeRangeQuery(datafeedConfig.getQuery(), timeField, start, end));
+
+        SearchRequest searchRequest = new SearchRequest(datafeedConfig.getIndices().toArray(new String[0])).source(searchSourceBuilder);
+        try (ThreadContext.StoredContext ignore = stashWithOrigin(client.threadPool().getThreadContext(), ML_ORIGIN)) {
+            SearchResponse response = client.execute(SearchAction.INSTANCE, searchRequest).actionGet();
+            List<? extends Histogram.Bucket> buckets = ((Histogram)response.getAggregations().get(DATE_BUCKETS)).getBuckets();
+            Map<Long, Long> hashMap = new HashMap<>(buckets.size());
+            for (Histogram.Bucket bucket : buckets) {
+                long bucketTime = toHistogramKeyToEpoch(bucket.getKey());
+                if (bucketTime < 0) {
+                        throw new IllegalStateException("Histogram key [" + bucket.getKey() + "] cannot be converted to a timestamp");
+                }
+                hashMap.put(bucketTime, bucket.getDocCount());
+            }
+            return hashMap;
+        }
+    }
+
+    private static long toHistogramKeyToEpoch(Object key) {
+        if (key instanceof DateTime) {
+            return ((DateTime)key).getMillis();
+        } else if (key instanceof Double) {
+            return ((Double)key).longValue();
+        } else if (key instanceof Long){
+            return (Long)key;
+        } else {
+            return -1L;
+        }
+    }
+
+    private static long calculateMissing(Map<Long, Long> indexedData, Bucket bucket) {
+        return indexedData.getOrDefault(bucket.getEpoch() * 1000, 0L) - bucket.getEventCount();
+    }
+
+    public static class BucketWithMissingData {
+
+        private final long missingData;
```

Comment:
`missingDocumentCount` would be more precise.  Also rename the argument and accessor to match.